### DISCUSSION TITLE
Branch Protection and Checks

### DISCUSSION THREAD

**Discussion by meza** [community/community#63427] [Pull Requests]
### Select Topic Area

Product Feedback

### Body

Over the past few weeks I've been noticing that the renovate PRs seem to get through even when the tests are failing so I went to investigate.

I found out that with the new updates to both GitHub and Renovate, I now need to enable the specific "Require status checks to pass before merging" option in the branch protection section. Fair enough I thought, I clicked and then moved on.

I still noticed PRs blasting through and breaking things, so I went to investigate more.

Imagine my horror when I found out that you now have to individually mark the checks that you want to enforce.
There is no option to say that "I want all checks to pass" before merging.
This is a horrible way to go about it and it should definitely not be the default behaviour.
Right now there are thousands of repositories where the owners are sleeping peacefully thinking that their automated setups protect them from bad dependency updates or other PRs and in reality, without their knowledge github just pulled the safety net out from under them.

Also, selecting the checks themselves is a horrible experience, especially when you have named checks for better readability. Like so: 
<img width="326" alt="image" src="https://github.com/community/community/assets/211977/8ea251fc-d5ad-4e76-9bb5-17d1f8a80986">

Without the option to say "All the checks should pass" the workflow now is to add a check in the yml, run it somewhere and then remember to go to the UI and hope that you can find it in the awkward dropdown.

I might be missing the point. It's totally possible. But this text below does not clarify anything.

<img width="567" alt="image" src="https://github.com/community/community/assets/211977/02c30c7c-bc1d-4f12-9997-65c0af9db6bc">

---

Here's my suggestion and ask:

- Change the default behaviour to "requrie all" and allow us to set up exclusions.
- On the off chance I'm missing the point and just by ticking the box that is actually what's happening, please add more clarification to the field

Thanks!

**Comment by HadrienG2** [comment]
The UX of those settings is truly the worst I've even been through in the GitHub UI by a large margin, and I strongly encourage you to review it. I generally agree with the above proposal of having a way to go for "require all + exceptions". But for completeness in case you decide to do it differently, here's my personal list of everything I hate about these settings:

- No shortcut for the "all the CI jobs" common case.
- Name of a matrixed job seems to work, but actually doesn't. Need to specify each matrix variant individually.
- No way to bulk-add matrix variants, must add them one by one.
- There is no clickable list, only a (slightly sluggish) search bar.
- Clicking an entry overrides the search bar contents, needs to reset it every time (thank god it doesn't scroll to the bottom of the list of required checks along the way).
- List of checks is sorted by order of insertion, with no way to reorder or re-sort it.
- Can't add a job unless it has run at least once. Queueing doesn't count.
    - If you vainly attempt to type in the name of the job in the search bar and press enter in an attempt to force-add it, it validates the current settings and exits the branch protection settings UI.
- If a job is renamed, the associated check stays required and blocks all PRs from being merged.
- Jobs that don't exist anymore stay forever as possible options in the search bar, making the handling of job renamings ever more confusing as the project ages.

**Comment by HonkingGoose** [comment]
The branch protection form is hard to understand. Here's another Renovate user getting confused by GitHub's text and layout:

- https://github.com/renovatebot/renovate/discussions/25750

If a full rewrite of the form is too difficult or too much work, I think the form can be improved by _moving_ the current content:

## Mockup

### Protect matching branches

- [ ] **Require a pull request before merging**
      When enabled, all commits must be made to a non-protected branch and submitted via a pull request before they can be merged into a branch that matches this rule. 

- [x] **Require status checks to pass before merging** 
       Choose which [status checks](https://docs.github.com/rest/commits/statuses) must pass before branches can be merged into a branch that matches this rule. When enabled, commits must first be pushed to another branch, then merged or pushed directly to a branch that matches this rule after status checks have passed.
       
  🔍 Search for status checks in the last week for this repository
  
  | Status checks that are required. |
  | --- |
  | Vercel |
  | Prettier |

  - [ ] **Require branches to be up to date before merging** 
        This ensures pull requests targeting a matching branch have been tested with the latest code. This setting will not take effect unless at least one status check is enabled (see below).
        
## Discussion

The layout of the form above is a bit better:

- The search box is _directly below_ the **Require status checks to pass before merging** text, so it's easier to understand that you must select a check
- The **Require branches to be up to date before merging** checkbox now comes _after_ the user has seen the **Require status checks to pass before merging** interface, so users are more likely to have selected a test already

Also, GitHub should validate the form properly:

- If a user selects **Require status checks to pass before merging** but does _not_ select any check: throw error message "You must select a check that should pass before merging!"
- If a user selects **Require branches to be up to date before merging** but does _not_ select any check: throw error message "You must select a check, before enabling Require branches to be up to date before merging"
